pkg/machines/components/storagePools/storagePoolVolumesTab.jsx

```diff
@@ -20,57 +20,127 @@
-                    const allocation = parseFloat(convertToUnit(volume.allocation, units.B, units.GiB).toFixed(2));
-                    const capacity = parseFloat(convertToUnit(volume.capacity, units.B, units.GiB).toFixed(2));
-                    const columns = [
-                        {
-                            name: (<div id={`${storagePoolIdPrefix}-volume-${volume.name}-name`}>{volume.name}</div>),
-                            header: true,
-                        }
-                    ];
-                    columns.push(
-                        {
-                            name: (<div id={`${storagePoolIdPrefix}-volume-${volume.name}-usedby`}>{usedBy[volume.name].join(', ')}</div>),
-                        }
-                    );
-                    columns.push(
-                        {
-                            name: (<div id={`${storagePoolIdPrefix}-volume-${volume.name}-size`}>{`${allocation} / ${capacity} GB`}</div>),
-                        }
-                    );
-
-                    return (<ListingRow columns={columns} key={`${storagePoolIdPrefix}-volume-${volume.name}`} />);
-                })}
-            </Listing>
-        </div>
-    );
-};
+    selectedChanged(volumeName, isSelected) {
+        let selected = [...this.state.selected];
+        if (!isSelected) {
+            let index = selected.indexOf(volumeName);
+            if (index !== -1)
+                selected.splice(index, 1);
+        } else
+            selected.push(volumeName);
+        this.setState({ selected });
+    }
+
+    resetSelection() {
+        this.setState({ selected: [] });
+    }
+
+    render() {
+        const { storagePool, vms } = this.props;
+        const storagePoolIdPrefix = storagePoolId(storagePool.name, storagePool.connectionName);
+        const columnTitles = [_("Name"), _("Used by"), _("Size")];
+        const volumes = storagePool.volumes || [];
+
+        if (volumes.length === 0) {
+            return (<div id={`${storagePoolIdPrefix}-storage-volumes-list`}>{_("No Storage Volumes defined for this Storage Pool")}</div>);
+        }
+
+        // Get a dictionary of vmName -> disks for a specific connection
+        const vmDisksMap = getVmDisksMap(vms, storagePool.connectionName);
+
+        // And make it a dictionary of volumeName -> array of Domains using volume
+        let isVolumeUsed = {};
+        for (let i in volumes) {
+            let volumeName = volumes[i].name;
+            isVolumeUsed[volumeName] = [];
+
+            for (let vmName in vmDisksMap) {
+                const disks = vmDisksMap[vmName];
+
+                for (let i in disks) {
+                    let disk = disks[i];
+                    if (disk.type == 'volume' && disk.volume == volumeName && disk.pool == storagePool.name)
+                        isVolumeUsed[volumeName].push(vmName);
+
+                    if (disk.type == 'file' && disk.source == [storagePool.path, volumeName.name].join('/'))
```

Comment:
`[storagePool.path, storageVolumeName].join('/')` is a constant within the outer `for` loop. Please move it there, to avoid this  extra step in the third-inner loop.

(Also, not too happy about an O(n³) here, but if the libvirt API doesn't allow a more efficient once, so be it :man_shrugging:)